### PR TITLE
Replace the repo contents with another independently developed implementation

### PR DESCRIPTION
**Pull request by generalmimon**
Based on https://github.com/Agile86/kaitai_rust/tree/rust_basic_support_v2 (permalink: [Agile86 / **kaitai_rust** > `a47ab47`](https://github.com/Agile86/kaitai_rust/tree/a47ab476efd7bbbd0abea95626901b5657b8c11d)). It brings the entire unchanged commit history into the repo (including commit hashes, so they're the same as in the original repos [Agile86/kaitai_rust](https://github.com/Agile86/kaitai_rust) and [bspeice/kaitai_rust](https://github.com/bspeice/kaitai_rust)), which reveals that the commits were authored by @bspeice, @Agile86 and @revitalyr (according to Git).

### CODE REVIEW

**Comment by generalmimon** [comment]
First of all, we need to solve one problem - no `LICENSE` file in the repo. As explained at https://choosealicense.com/no-permission/:

> When you make a creative work (which includes code), the work is under exclusive copyright by default. Unless you include a license that specifies otherwise, nobody else can copy, distribute, or modify your work without being at risk of take-downs, shake-downs, or litigation. Once the work has other contributors (each a copyright holder), “nobody” starts including you.

Therefore, it is desirable to choose a license. If you're mentioned, **please comment whether is it OK to release your changes under the [MIT license](https://choosealicense.com/licenses/mit/)**.

@bspeice, I assume you own the code in your https://github.com/bspeice/kaitai_rust repo, so your consent with the proposed [MIT](https://choosealicense.com/licenses/mit/) license should be enough. @Agile86 and @revitalyr, please check with your employer whether they agree with releasing your work in https://github.com/Agile86/kaitai_rust/tree/rust_basic_support_v2 under [MIT](https://choosealicense.com/licenses/mit/).

**Comment by sf-jonstewart** [comment]
@Agile86 and @revitalyr, you have permission to release this under the MIT license. Thank you for your hard work!

**Comment by Agile86** [comment]
I agree to publish the code under the [MIT license](https://choosealicense.com/licenses/mit/).

**Comment by revitalyr** [comment]
I agree to publish the code under the [MIT license](https://choosealicense.com/licenses/mit/).

**Comment by generalmimon** [comment]
@Agile86 @revitalyr Thank you for your approval!

**Comment by bspeice** [comment]
The license is fine by me as well.

**Comment by generalmimon** [comment]
@bspeice Thanks!

File: src/lib.rs [hunk 1]
```diff
@@ -1,203 +1,960 @@
...
+        }
     }
+}
 
-    #[test]
-    fn read_f4be() {
-        let mut buf = Cursor::new(vec![62, 128, 0, 0]);
-        assert_eq!(buf.read_f4be().unwrap(), 0.25);
+pub trait KStream {
+    fn clone(&self) -> BytesReader;
```
Comment (generalmimon) [inline src/lib.rs:268]: I don't really understand the purpose of this trait. In theory, it makes sense to allow alternative implementations of `KStream`, but the generated code can only work with `BytesReader`. In fact, the generated code only uses this trait to call the `clone()` method, which is required here to return `BytesReader`.

Unfortunately there are such illogical things in your code everywhere... (and this one is still relatively benign compared to others). I don't know if we can ever fix them all.

**Pull request merged**